README.md
# Parrot Sample
* This project is developed on `Python Flask` and is used to support sampling, demonstration of interface automation/traffic playback tools.
* Currently only basic HTTP operations are provided, data is stored in running memory, will be reset after restarting the service.
* Your issues and participates in the construction are welcome.

## User Handbook
### Step 0: Basic environment preparation
* This project is based on python 2.7.x or 3.x. Please make sure that `python` and `pip` are installed on the machine running this service. Recommended version: 3.7.x
* The dependent modules have been written in requirements.txt and can be installed using `pip install -r requirements.txt`

### Step 1: Boot the application
* This project uses port 8080 by default. If there is a conflict, you can modify `_PORT_` in `app.py`.
* Start the application in command line mode: `python app.py`, if you need to debug, add `debug=True` parameter to app.run.
* If it starts normally, the screen will display "Running on http://0.0.0.0:8080/" content output, WARNING can be ignored.
* If it does not start normally, please confirm Step 0 and port usage. If you have any questions, please feedback Issue.

### Step 2: Visit the application
* The server ip can be obtained by `ipconfig` or `ifconfig` command, and then spliced ​​into a complete address, such as: `10.100.100.10:8080`, 127.0.0.1 is not recommended.
* The above address can be directly accessed in the browser window. If it is normal, the page function module will be presented.
* This site is not compatible with mobile styles, and the mobile display is not effective.
* If the access is abnormal, please make sure that the application of Step 1 is running normally, and then there is a problem to feedback Issue.

### What the application supports now
* Basic HTTP GET/POST requests. Except for the index page, the interface response text is in json format, including `timestamp` and random `tag`.
* Random exception rate, longest and shortest random time-consuming ranges, can be edited online.

#### User's Behavior
* Register: `POST` method
* Unregister: `POST` method, `token` is added in `headers`
* Logon: `POST` method, `token` would be generated and set in response `headers` and `cookies`
* Logout: `POST` method, `token` is added in `headers`, `cookies` would be cleared

#### User's Hobbies
* Hobby List: `GET` method, `token` is added in `headers`
* Hobby Detail：`GET` method, `token` is added in `headers`, and `name` parameter should be obtained from response of `Hobby List`
* Add a Hobby: `POST` method, `token` is added in `headers`
* Remove a Hobby: `POST` method, `token` is added in `headers`
* Random Suggestion: `GET` method, `token` is added in `headers`, and a `today()` parameter is added

#### Suggested demo operations
* Logon => Hobby List => Add a Hobby => Hobby List => Hobby Detail => Random Suggestion => Logout
* You can view the corresponding interface call details in the developer tool of the browser, or you can export the HAR file.


CHANGELOG.md
What's new in Sample 1.1?
===========================
Release Date: 2019-09-26

Release Feature:
 * Add English page, Chinese and English pages can be switched back and forth
 * Config could be edited online


What's new in Sample 1.0?
===========================
Release Date: 2019-09-17

Release Feature:
 * Rebuild `index.html` and `app.py`
 * Supports HTTP GET and POST requests, headers and cookies, token validation, parameter dependencies, etc.


What's new in Sample 0.2?
===========================
Release Date: 2019-03-30

Release Feature:
 * Add a random interface


What's new in Sample 0.1?
===========================
Release Date: 2019-03-19

Release Feature:
 * Support Simple/Complex http GET/POST requests

module/helper.py
# -*- coding: utf-8 -*-

import os
import sys
import time
import datetime
import shutil
import random
import string
import re
import json
import socket


"""
Time Helper
"""

    
def today(form='%Y-%m-%d'):
    return datetime.datetime.now().strftime(form)


def days_ago(days=0, form='%Y-%m-%d'):
    return (datetime.datetime.now() - datetime.timedelta(days=days)).strftime(form)


def days_later(days=0, form='%Y-%m-%d'):
    return (datetime.datetime.now() + datetime.timedelta(days=days)).strftime(form)


def now(form='%Y-%m-%d %H:%M:%S'):
    return datetime.datetime.now().strftime(form)


def now_ms(form='%Y-%m-%d %H:%M:%S'):  # e.g. 2016-01-01 01:01:01.012
    _now = datetime.datetime.now()
    return _now.strftime(form) + '.' + "%03d" % round(float(_now.microsecond) / 1000)


def now_timestamp():  # e.g. 1451581261
    return int(round(time.time()))


def now_timestamp_ms():  # e.g. 1451581261339
    _now = datetime.datetime.now()
    return int(time.mktime(
        time.strptime(_now.strftime('%Y-%m-%d %H:%M:%S'), '%Y-%m-%d %H:%M:%S'))) * 1000 + _now.microsecond / 1000


def hours_ago(hours=0, form='%Y-%m-%d %H:%M:%S'):
    return (datetime.datetime.now() - datetime.timedelta(hours=hours)).strftime(form)


def hours_later(hours=0, form='%Y-%m-%d %H:%M:%S'):
    return (datetime.datetime.now() + datetime.timedelta(hours=hours)).strftime(form)


"""
File Helper
"""


def get_project_path():
    _r = re.compile(r"(^.+[/\\])(.*)")
    _p = os.path.abspath(__file__)
    if os.path.exists("{}/.iparrot".format(_p)):
        return _p
    while re.search(_r, _p):
        _p = re.findall(_r, _p)[0][0]
        if os.path.exists("{}/.iparrot".format(_p)):
            return _p
        else:
            _p = os.path.abspath(_p)
    return os.path.abspath(__file__)


def get_file_name(file, ext=0):
    file = str(file).split('/')[-1].split('\\')[-1]
    return file if ext else file.replace(".{}".format(file.split('.')[-1]), '')


def get_file_path(file):
    return os.path.dirname(file)


def make_dir(directory):
    if not os.path.exists(directory):
        try:
            os.makedirs(directory)
        except OSError:
            return False

    return directory


def remove_dir(directory):
    if os.path.exists(directory):
        try:
            if os.path.isfile(directory):
                os.remove(directory)
            elif os.path.isdir(directory):
                shutil.rmtree(directory)
        except OSError:
            return False

    return directory


def remove_file(filename):
    if os.path.exists(filename):
        try:
            if os.path.isfile(filename):
                os.remove(filename)
            elif os.path.isdir(filename):
                shutil.rmtree(filename)
        except OSError:
            return False

    return filename


def copy_file(source, target):
    try:
        if os.path.isdir(source):
            shutil.copytree(source, target)
        else:
            shutil.copy(source, target)
    except OSError:
        return False

    return target


def get_dir_folders(directory):
    if os.path.isdir(directory):
        return [name for name in os.listdir(directory) if os.path.isdir(os.path.join(directory, name))]
    else:
        return []


def get_dir_files(directory):
    if os.path.isdir(directory):
        return ["{}/{}".format(directory, name) for name in os.listdir(directory) if os.path.isfile(os.path.join(directory, name))]
    else:
        return []


"""
Random Helper
"""


def get_random_integer(length=10, head=None, tail=None):
    """
    :param length: an integer
    :param head: to be started with
    :param tail: to be ended with
    :return: a random integer
    """
    my_str = random.randint(10 ** (length-1), 10 ** length - 1)
    if head:
        head = str(head)
        if len(head) >= length:
            return head
        my_str = int(head + str(str(my_str)[len(head):]))
    if tail:
        tail = str(tail)
        if len(tail) >= length:
            return tail
        my_str = int(str(str(my_str)[:-len(tail)]) + tail)

    return my_str


def get_random_string(length=10, simple=1, head=None, tail=None):
    """
    :param length: an integer
    :param simple:
    :param head: to be started with
    :param tail: to be ended with
    :return: a random string
    """
    if simple == 1:  # base: digits and letters
        src = str(string.digits) + string.ascii_letters
    elif simple == 2:  # on base, remove confusing letters: 0&O&o c&C I&l&1 k&K p&P s&S v&V w&W x&X z&Z
        src = 'abdefghijmnqrtuyABDEFGHJLMNQRTUY23456789'
    else:  # on base, add some special letters
        src = str(string.digits) + string.ascii_letters + '~!@#$%^&*,.-_=+'

    # in case the length is too high
    src = src * int((length / len(src) + 1))

    my_str = ''.join(random.sample(src, length))

    if head:
        head = str(head)
        if len(head) >= length:
            return head
        my_str = head + str(my_str[len(head):])
    if tail:
        tail = str(tail)
        if len(tail) >= length:
            return tail
        my_str = str(my_str[:-len(tail)]) + tail

    return str(my_str)


def get_random_phone(head=None, tail=None):
    """
    :param head: to be started with
    :param tail: to be ended with
    :return: a random phone number in China
    """
    length = 11
    my_str = '1' + "".join(random.choice('3456789')) + "".join(random.choice(string.digits) for i in range(length-2))

    if head:
        head = str(head)
        if len(head) >= length:
            return head
        my_str = head + str(my_str[len(head):])
    if tail:
        tail = str(tail)
        if len(tail) >= length:
            return tail
        my_str = str(my_str[:-len(tail)]) + tail

    return my_str


"""
Dict Helper
"""


# Traverse the dictionary/list to get all key/value pairs with depth of 1
def get_all_kv_pairs(item, prefix=None, mode=1):
    """
    :param item: a dict or list, with any depth
    :param prefix: prefix of a key, used when recursive
    :param mode: 1: depth == 1, 0: extract keys of all levels
    :return: all key/value pairs with depth of 1
    """
    _pairs = {}
    if not mode and prefix:
        _pairs[prefix] = item
    if not item and prefix:
        _pairs[prefix] = item
    if isinstance(item, dict):  # A.B
        for _key, _val in item.items():
            _key = "{}.{}".format(prefix, _key) if prefix else _key
            if not mode:
                _pairs[_key] = _val
            _pairs.update(get_all_kv_pairs(item=_val, prefix=_key, mode=mode))
    elif isinstance(item, (list, set, tuple)):  # A[0]
        for i, _key in enumerate(item):
            _key = "{}[{}]".format(prefix, i) if prefix else "[{}]".format(i)
            if not mode:
                _pairs[_key] = item[i]
            _pairs.update(get_all_kv_pairs(item=item[i], prefix=_key, mode=mode))
    else:  # the end level
        if prefix:
            _pairs[prefix] = item
        else:
            _pairs[item] = ''
    return _pairs


def get_matched_keys(key, keys, fuzzy=1):
    """
    :param key: a node or a list of nodes, which would be matched with
    :param keys: a list of nodes, which would be matched
    :param fuzzy: 1-fuzzy match, 0-exact match
    :return: a list of matched keys
    """
    descendant_keys = []
    if not isinstance(key, (list, set, tuple)):
        key = [str(key), ]
    if not isinstance(keys, (list, set, tuple)):
        keys = [str(keys), ]
    if not key:
        return keys
    for i_key in key:
        for j_key in keys:
            if fuzzy:
                if i_key in j_key:
                    descendant_keys.append(j_key)
            else:
                if i_key == j_key:
                    descendant_keys.append(j_key)
                elif j_key.startswith("{}.".format(i_key)):
                    descendant_keys.append(j_key)
                elif j_key.startswith("{}[".format(i_key)):
                    descendant_keys.append(j_key)
    return descendant_keys


# parameters to dict, e.g. date=0305&tag=abcd&id=123 => {'date': '0305', 'tag': 'abcd', 'id': 123}
def param2dict(item):
    my_dict = {}
    for i_item in str(item).split('&'):
        i_arr = i_item.split('=')
        my_dict[i_arr[0]] = i_arr[1]
    return my_dict


# dict to parameters, e.g. {'date': '0305', 'tag': 'abcd', 'id': 123} => date=0305&tag=abcd&id=123
def dict2param(item):
    if not isinstance(item, dict):
        return str(item)
    my_param = ''
    for key, val in item.items():
        my_param = my_param + "&{}={}".format(key, val) if my_param else "{}={}".format(key, val)
    return my_param


def get_host_ip():
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(('8.8.8.8', 80))
        ip = s.getsockname()[0]
    finally:
        s.close()

    return ip


if __name__ == "__main__":
    har_time2timestamp(har_time='aaaa')
    print(get_project_path())
    resp = {
        "status.code": 200,
        "content": {
            "error": "",
            "status": 200,
            "usedtime": 0,
            "srvtime": "2019-8-22 16:48:35",
            "islogin": 0,
            "iswork": True,
            "tag": "3301c39cc5fc4448a3a325c231cfa475",
            "data": {
                "msgCode": "1000",
                "msgInfo": "登录成功",
                "data": "",
                "RequestId": "e1e527d2-2a02-476c-954c-cf1bf0e61159",
                "firstloginCoupon": "",
                "MobileStatus": "1",
                "MobileNo": "",
                "LoginType": 1
            }
        },
        "headers": {
            "Date": "Thu, 22 Aug 2019 08:48:34 GMT",
            "Content-Type": "application/json; charset=utf-8",
            "Content-Length": "323",
            "Connection": "keep-alive",
            "Server": "openresty/1.11.2.1",
            "Cache-Control": "private",
            "X-AspNetMvc-Version": "4.0",
            "Access-Control-Allow-Origin": "",
            "X-AspNet-Version": "4.0.30319",
            "X-Powered-By": "ASP.NET",
            "id": "TCWEBV088154",
            "Leonid-Ins": "735365034-HitConf-2[m]-59e85433a1a367bed390c711",
            "p3p": "CP=\"IDC DSP COR ADM DEVi TAIi PSA PSD IVAi IVDi CONi HIS OUR IND CNT\"",
            "Leonid-addr": "NTguMjExLjExMS4xOA==",
            "Leonid-Time": "0"
        },
        "cookies": {}
    }
    print(json.dumps(get_all_kv_pairs(item=resp, mode=0)))



app.py
# -*- coding: utf-8 -*-

from flask import Flask
from flask import render_template
from flask import request
from flask import redirect
from flask import url_for
from flask import make_response
import hashlib

from module.helper import *


app = Flask(__name__)

_IP_ = get_host_ip()
_PORT_ = '8080'

CONFIG = {
    'EXCEPTION_PERCENT': 0,  # 0-100
    'DURATION_MAX': 1000,  # ms
    'DURATION_MIN': 50,  # ms
}

msg_ok_default = {
    'code': 100,
    'message': "OK."
}
msg_error_default = {
    'code': 500,
    'message': "The system is busy, please try again later."
}

users = {
    'sample': {
        'password': "a10d74a680a1c5cf5bfe6cbbaec1d2e1",  # by md5_password()
        'token': "",
        'hobbies': [
            {
                'name': 'reading',
                'frequency': "1 book per month"
            },
            {
                'name': 'running',
                'frequency': "10 km per week"
            }
        ]
    }
}


@app.route('/')
def blank():
    return redirect(url_for('index'))


@app.route('/index')
def index():
    return render_template('index_en.html',
                           base_url="http://{}:{}/".format(_IP_, _PORT_),
                           conf_rate=CONFIG['EXCEPTION_PERCENT'],
                           conf_max_time=CONFIG['DURATION_MAX'],
                           conf_min_time=CONFIG['DURATION_MIN']
                           )


@app.route('/zh_cn')
def index_en():
    return render_template('index.html',
                           base_url="http://{}:{}/".format(_IP_, _PORT_),
                           conf_rate=CONFIG['EXCEPTION_PERCENT'],
                           conf_max_time=CONFIG['DURATION_MAX'],
                           conf_min_time=CONFIG['DURATION_MIN']
                           )


@app.route('/hello')
def hello():
    return make_response({'code': 100, 'msg': 'Hello Parrot Sample.'})


@app.route('/register', methods=['GET', 'POST'])
def register():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    if request.method == 'GET':
        username = request.args.get('username', None)
        password = request.args.get('password', None)
    else:
        username = request.form.get('username', None)
        password = request.form.get('password', None)

    msg_invalid_input = {
        'code': 101,
        'message': "Invalid username or password, please modify and try again."
    }
    msg_user_exists = {
        'code': 102,
        'message': "User already exists in system, please use another username."
    }

    if not username or not password:
        rsp = make_response(msg_invalid_input)
    elif username in users.keys():
        rsp, _ = random_response(msg_user_exists)
    else:
        rsp, flag = random_response()
        if flag:
            users[username] = {
                'password': md5_password(password),
                'token': "",
                'hobbies': []
            }
    return rsp


@app.route('/unregister', methods=['GET', 'POST'])
def unregister():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    username, rsp = is_logged_out()
    if rsp:
        return rsp
    rsp, flag = random_response()
    if flag:
        if username in users.keys():
            del users[username]
        rsp.delete_cookie('username')
        rsp.delete_cookie('token')
    return rsp


@app.route('/logon', methods=['GET', 'POST'])
def logon():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    if request.method == 'GET':
        username = request.args.get('username', None)
        password = request.args.get('password', None)
    else:
        username = request.form.get('username', None)
        password = request.form.get('password', None)

    msg_invalid_input = {
        'code': 201,
        'message': "The username or password is incorrect, please modify and log in again."
    }

    if not username or not password or \
            username not in users.keys() or md5_password(password) != users[username]['password']:
        return make_response(msg_invalid_input)
    rsp, flag = random_response()
    if flag:
        token = generate_token(username)
        rsp.headers['token'] = users[username]['token'] = token
        rsp.set_cookie('username', username)
        rsp.set_cookie('token', token)
    return rsp


@app.route('/logout', methods=['GET', 'POST'])
def logout():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    rsp, flag = random_response()
    if flag:
        username = request.cookies.get('username', None)
        if username:
            users[username]['token'] = ""
            rsp.delete_cookie('username')
            rsp.delete_cookie('token')
    return rsp


@app.route('/my_hobby', methods=['GET'])
@app.route('/my_hobbies', methods=['GET'])
@app.route('/hobby_list', methods=['GET'])
def hobby_list():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    username, rsp = is_logged_out()
    if rsp:
        return rsp

    rsp, _ = random_response(dict(msg_ok_default, **{'hobbies': users[username]['hobbies']}))
    return rsp


@app.route('/hobby_detail', methods=['GET'])
def hobby_detail():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    username, rsp = is_logged_out()
    if rsp:
        return rsp

    name = request.args.get('name', None)
    if not name:
        return random_response(
            {'code': 301, 'message': "Invalid input, 'name' argument is required."}
        )[0]
    for hobby in users[username]['hobbies']:
        if name == hobby['name']:
            return random_response(
                dict(msg_ok_default, **{'detail': hobby})
            )[0]
    return random_response(
        {'code': 302, 'message': "Invalid input, the specified hobby name doesn't exist."}
    )[0]


@app.route('/add_hobby', methods=['GET', 'POST'])
@app.route('/hobby_add', methods=['GET', 'POST'])
def hobby_add():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    username, rsp = is_logged_out()
    if rsp:
        return rsp

    if request.method == 'GET':
        name = request.args.get('name', None)
        frequency = request.args.get('frequency', None)
    else:
        name = request.form.get('name', None)
        frequency = request.form.get('frequency', None)

    if not name or not frequency:
        return random_response(
            {'code': 311, 'message': "Invalid input, 'name' and 'frequency' arguments are needed."}
        )[0]

    for hobby in users[username]['hobbies']:
        if name == hobby['name']:
            hobby['frequency'] = frequency
            return random_response(
                dict(msg_ok_default, **{'hobbies': users[username]['hobbies']})
            )[0]
    users[username]['hobbies'].append({
        'name': name,
        'frequency': frequency
    })
    return random_response(
        dict(msg_ok_default, **{'hobbies': users[username]['hobbies']})
    )[0]


@app.route('/remove_hobby', methods=['GET', 'POST'])
@app.route('/hobby_remove', methods=['GET', 'POST'])
def hobby_remove():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    username, rsp = is_logged_out()
    if rsp:
        return rsp

    if request.method == 'GET':
        name = request.args.get('name')
    else:
        name = request.form.get('name')

    if name:
        for idx, hobby in enumerate(users[username]['hobbies']):
            if name == hobby['name']:
                del users[username]['hobbies'][idx]
                break
    else:
        if len(users[username]['hobbies']):
            del users[username]['hobbies'][0]
    return random_response(
        dict(msg_ok_default, **{'hobbies': users[username]['hobbies']})
    )[0]


@app.route('/suggest_hobby', methods=['GET'])
@app.route('/hobby_suggest', methods=['GET'])
def hobby_suggest():
    add_timestamp(msg_ok_default)
    add_tag(msg_ok_default)

    username, rsp = is_logged_out()
    if rsp:
        return rsp

    _today = request.args.get('today', now())
    if len(users[username]['hobbies']) <= 1:
        sug_msg = {
            'code': 303,
            'message': "You need to add more hobbies firstly."
        }
    else:
        idx = (int(_today[-1]) + random.randint(0, len(users[username]['hobbies'])-1)) % len(users[username]['hobbies'])
        suggest = users[username]['hobbies'][idx]['name']
        sug_msg = {
            'code': 100,
            'message': "The suggested hobby for today is: {}".format(suggest)
        }
    return random_response(sug_msg)[0]


@app.route('/update_config', methods=['GET', 'POST'])
def update_config():
    if request.method == 'GET':
        _rate = request.args.get('rate')
        _max = request.args.get('max')
        _min = request.args.get('min')
    else:
        _rate = request.form.get('rate')
        _max = request.form.get('max')
        _min = request.form.get('min')

    try:
        _rate = int(_rate)
        _max = int(_max)
        _min = int(_min)
        if _max < _min:
            _max, _min = _min, _max
        CONFIG['EXCEPTION_PERCENT'] = _rate
        CONFIG['DURATION_MAX'] = _max
        CONFIG['DURATION_MIN'] = _min
        return make_response(
            {
                'code': 900,
                'message': "OK.",
                'config': CONFIG
            })
    except ValueError:
        return make_response(
            {
                'code': 901,
                'message': "Invalid input. No change.",
                'config': CONFIG
            })


def random_response(ok_msg=msg_ok_default):
    time.sleep(random.randint(CONFIG['DURATION_MIN'], CONFIG['DURATION_MAX'])*1.0/1000)
    if int(CONFIG['EXCEPTION_PERCENT']) >= random.randint(1, 100):
        rsp = make_response(msg_error_default)
        rsp._status = "500 Internal Server Error"
        return rsp, False
    else:
        return make_response(ok_msg), True


def generate_token(user):
    hl = hashlib.md5()
    hl.update("{}_{}".format(user, now_timestamp()).encode('utf-8'))
    return hl.hexdigest()


def md5_password(password, key='Sample'):
    hl = hashlib.md5()
    hl.update("{}_{}".format(key, password).encode('utf-8'))
    return hl.hexdigest()


def is_logged_out():
    username = request.cookies.get('username', None)
    c_token = request.cookies.get('token', None)
    r_token = request.headers.get('token', None)
    msg_logged_out = {
        'code': 211,
        'message': "Login token expired, please log in again."
    }
    msg_logged_by_other = {
        'code': 212,
        'message': "Account logged in by others, please log in again."
    }
    # print("{} {} {}".format(username, c_token, r_token))
    if not r_token or r_token == 'null':
        return None, make_response(msg_logged_out)
    elif r_token != users[username]['token']:
        return None, make_response(msg_logged_by_other)
    return username, None


def add_timestamp(msg):
    msg['timestamp'] = now_timestamp()


def add_tag(msg):
    msg['tag'] = get_random_string(32)


if __name__ == '__main__':
    app.run(host='0.0.0.0', port=_PORT_)
